Modify S1117: Migrate To LayC - shadow variables (#3270)
SonarSource/rspec: rules/S1117/javascript/rule.adoc

include::../why-general.adoc[]

> Note that variables can also shadow functions.

Reviewer: 👍

> Note that variables can also shadow functions.

Reviewer: ```suggestion
Note that functions in JavaScript are first-class citizens. This means that they possess the same attributes as variables, including the ability to shadow other variables and, conversely, be shadowed by them. 
```

=== Noncompliant code example

The example below shows the typical situations in which shadowing can occur.

[source,javascript]
----

> function example(items) {

Reviewer: ```suggestion
function outer(items) {
```

  var counter = 0;

>   let nested = function(items) { // Noncompliant: the parameter "items" is shadowed.

Reviewer: ```suggestion
  function inner(items) { // Noncompliant: the parameter "items" is shadowed.
```

    var counter = counter + 1; // Noncompliant: the outer "counter" is shadowed.
  }

>   nested(items);

Reviewer: ```suggestion
  inner(items);
```
